temporalio/worker/workflow_sandbox/importer.py

```diff
@@ -0,0 +1,287 @@
+
+        # Set our modules, then unset on complete
+        sys.modules = self.new_modules
+        builtins.__import__ = self.import_func
+        try:
+            # The first time this is done, we need to manually import a few
+            # implied modules
+            if "sys" not in self.new_modules:
+                # Refresh sys
+                self.new_modules["sys"] = importlib.__import__("sys")
+                # Import builtins and replace import
+                self.new_modules["builtins"] = importlib.__import__("builtins")
+                self.new_modules["builtins"].__import__ = self.import_func  # type: ignore
+
+                # Create a blank __main__ to help with anyone trying to
+                # "import __main__" to not accidentally get the real main
+                self.new_modules["__main__"] = types.ModuleType("__main__")
+                # Reset sys modules again
+                sys.modules = self.new_modules
+
+            # Restrict builtins
+            self.env.restrict_builtins()
+
+            yield None
+        finally:
+            sys.modules = self.orig_modules
+            builtins.__import__ = self.orig_import
+            self.env.unrestrict_builtins()
+
+    @contextmanager
+    def _unapplied(self) -> Iterator[None]:
+        # Error if already unapplied
+        if sys.modules is not self.new_modules:
+            raise RuntimeError("Sandbox importer already unapplied")
+
+        # Set orig modules, then unset on complete
+        sys.modules = self.orig_modules
+        builtins.__import__ = self.orig_import
+        self.env.unrestrict_builtins()
+        try:
+            yield None
+        finally:
+            sys.modules = self.new_modules
+            builtins.__import__ = self.import_func
+            self.env.restrict_builtins()
+
+    def _traced_import(
+        self,
+        name: str,
+        globals: Optional[Mapping[str, object]] = None,
+        locals: Optional[Mapping[str, object]] = None,
+        fromlist: Sequence[str] = (),
+        level: int = 0,
+    ) -> types.ModuleType:
+        _trace("Importing %s (fromlist: %s, level: %s)", name, fromlist, level)
+        global _trace_depth
+        _trace_depth += 1
+        try:
+            return self._import(name, globals, locals, fromlist, level)
+        finally:
+            _trace_depth -= 1
+
+    def _import(
+        self,
+        name: str,
+        globals: Optional[Mapping[str, object]] = None,
+        locals: Optional[Mapping[str, object]] = None,
+        fromlist: Sequence[str] = (),
+        level: int = 0,
+    ) -> types.ModuleType:
+        # Check module restrictions and passthrough modules
+        if name not in sys.modules:
+            # Make sure not an entirely invalid module
+            self.env.assert_valid_module(name)
+
+            # Check if passthrough
+            with self._unapplied():
+                passthrough_mod = self.env.maybe_passthrough_module(name)
```

Comment:
nit: we trigger the `_unapplied()` context manager on each import even when the module is not passthrough, looks like you can check if it's passthrough before entering the unapplied context.

Reply:
I didn't want to split up `maybe_passthrough_module` to two steps if I didn't have to (one for check another for actually obtaining from unapplied sys modules).

This code is actually the result of me expecting to not have global sys.module replacement. I just ended up having to do that for global leaking reasons I could not solve in time. In theory maybe_passthrough_module was meant to run outside the sandbox.

But I agree this can be improved. This is all just time based (you should have seen it before).

Reply:
Now we only unapply after we've matched